src/ImageSharp/Formats/Qoi/QoiEncoderCore.cs

```diff
@@ -0,0 +1,222 @@
+
+/// <summary>
+/// Image encoder for writing an image to a stream as a QOi image
+/// </summary>
+public class QoiEncoderCore : IImageEncoderInternals
+{
+    /// <summary>
+    /// The encoder with options
+    /// </summary>
+    private readonly QoiEncoder encoder;
+
+    /// <summary>
+    /// Used the manage memory allocations.
+    /// </summary>
+    private readonly MemoryAllocator memoryAllocator;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="QoiEncoderCore"/> class.
+    /// </summary>
+    /// <param name="encoder">The encoder with options.</param>
+    /// <param name="memoryAllocator">The <see cref="MemoryAllocator" /> to use for buffer allocations.</param>
+    public QoiEncoderCore(QoiEncoder encoder, MemoryAllocator memoryAllocator)
+    {
+        this.encoder = encoder;
+        this.memoryAllocator = memoryAllocator;
+    }
+
+    /// <inheritdoc />
+    public void Encode<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Guard.NotNull(image, nameof(image));
+        Guard.NotNull(stream, nameof(stream));
+
+        this.WriteHeader(image, stream);
+        this.WritePixels(image, stream);
+        WriteEndOfStream(stream);
+        stream.Flush();
+    }
+
+    private void WriteHeader(Image image, Stream stream)
+    {
+        // Get metadata
+        Span<byte> width = stackalloc byte[4];
+        Span<byte> height = stackalloc byte[4];
+        BinaryPrimitives.WriteUInt32BigEndian(width, (uint)image.Width);
+        BinaryPrimitives.WriteUInt32BigEndian(height, (uint)image.Height);
+        QoiChannels qoiChannels = this.encoder.Channels ?? QoiChannels.Rgba;
+        QoiColorSpace qoiColorSpace = this.encoder.ColorSpace ?? QoiColorSpace.SrgbWithLinearAlpha;
+
+        // Write header to the stream
+        stream.Write(QoiConstants.Magic);
+        stream.Write(width);
+        stream.Write(height);
+        stream.WriteByte((byte)qoiChannels);
+        stream.WriteByte((byte)qoiColorSpace);
+    }
+
+    private void WritePixels<TPixel>(Image<TPixel> image, Stream stream)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        // Start image encoding
+        using IMemoryOwner<Rgba32> previouslySeenPixelsBuffer = this.memoryAllocator.Allocate<Rgba32>(64, AllocationOptions.Clean);
+        Span<Rgba32> previouslySeenPixels = previouslySeenPixelsBuffer.GetSpan();
+        Rgba32 previousPixel = new(0, 0, 0, 255);
+        Rgba32 currentRgba32 = default;
+        Buffer2D<TPixel> pixels = image.Frames[0].PixelBuffer;
+
+        for (int i = 0; i < pixels.Height; i++)
+        {
+            Span<TPixel> row = pixels.DangerousGetRowSpan(i);
+            for (int j = 0; j < row.Length && i < pixels.Height; j++)
+            {
+                // We get the RGBA value from pixels
+                TPixel currentPixel = row[j];
+                currentPixel.ToRgba32(ref currentRgba32);
```

Comment:
These conversions can be done much faster in batches thanks to SIMD. A pattern we commonly do is to convert the entire image row into a temporary working buffer. You would need to do something like:

```C#
        using IMemoryOwner<Rgba32> rgbaRowBuffer = this.memoryAllocator.Allocate<Rgba32>(pixels.Width);
        Span<Rgba32> rgbaRow = rgbaRowBuffer.GetSpan();

        for (int i = 0; i < pixels.Height; i++)
        {
            Span<TPixel> row = pixels.DangerousGetRowSpan(i);
            PixelOperations<TPixel>.Instance.ToRgba32(this.PASS_CONFIGURATION_TO_THE_ENCODER, row, rgbaRow);
            for (int j = 0; j < row.Length && i < pixels.Height; j++)
            {
                currentRgba32 = rgbaRow[j];
```

And also update `rgbaRow` in the QOI_OP_RUN.